### DISCUSSION TITLE
Displaying supported sensors in README.md

### DISCUSSION THREAD

**Discussion by Thrilleratplay** [custom-components/ble_monitor#439] [Ideas]
When adding the Govee devices to README.md, I found adding the line " Govee devices only send data with the option active_scan set to True, so make sure you change this setting, as the default is False" to every device was redundant.  What started off as a change to add a column to the Supported Sensors table turned into an attempt to "simplify" the table. 

[This did not turn out as well as I hoped](https://github.com/Home-Is-Where-You-Hang-Your-Hack/ble_monitor/tree/README_refactor#supported-sensors).  An attempt to group like devices by shared functionality became confusing as there are enough differences (example: does an alarm clock that transmits temperature and humidity go under Thermometers/Hydrometers?).  Additionally many devices are manufactured under different brands with and without variations so grouping by brand cannot be done.  Instead it was left as one table, but close to finishing but see a table that is more complicated than when I started.

Perhaps I am over thinking it and am solving a problem that does not exist.  If it is an issue, the only thing I can think of is to leave a list of sensors in the README with a link to specific device details in another markdown file.  This however may lead to information being missed.  

Thoughts?  Concerns?  Ridicule?

**Comment by Ernst79** [comment]
Yes, I totally understand the struggle you have with the table, I have the exact same thing and  the same thoughts have crossed my mind how to make it better.

the table, of course, has a history, it started small with a few sensors, which was readable. But now it has grown till what it is now.

one possibility is grouping per advertisement format, so make a different table for qingping, xiaomi MiBeacon, govee, etc. 

another possibility that crossed my mind is splitting up the entire readme, with an individual page for
- introduction + installation
- Supported sensors
- Device tracking
- config at integration level
- Config at device level
- Faq

the introduction could have a short list with only the supported sensor types with a link to the second page. each page should have a link to the other pages.

An additional idea is to create a different landing page for hacs, that refers to the pages on github. My experience is that the links in Hacs never work, as the are relative links. This would also have the advantage of not having to update the same things twice. For hacs we use the info.md page, which is basically a copy of the README.md with slightly changed links for pictures, to make them work.

Personally, I have the best feeling with the second option, multiple pages

**Reply by Thrilleratplay** [reply]
@Ernst79  I have to agree.  The second option also crossed my mind; a few months ago I updated the Jekyll documentation for Heads ROM wiki ([site](https://osresearch.net/), [repo](https://github.com/osresearch/heads-wiki/)) and thought searching/tagging would help break down the information for each sensor.  This also did not seem practical as it would not help the situation for `info.md`.

If it isn't too assuming of me, I will experiment with rearranging the documentation for a github pages site.

**Comment by Ernst79** [comment]
Sure, that would be more than welcome. You still combine it with the first option, on the sensor page, you can still split up in sensors per advertisement format or something like that. 

Just to let you know, I’m currently working on writing unittests. I might ask you to send me some advertisements for govee sensors, to write the tests, but I will first create the basic setup.

**Reply by Ernst79** [reply]
Just looked at the link you had in your post, something like that also looks nice. I’ll let you experiment

**Reply by Thrilleratplay** [reply]
@Ernst79 There is still quite a bit of polishing still to do but wanted to make sure this would work for the project.  


https://home-is-where-you-hang-your-hack.github.io/ble_monitor/

A few notes:
 * I have not pulled in new devices from the past two weeks yet
 * ~existing links have not been updated within the pages and the menu will be required to navigate between pages.~

The main difference is now [devices have been split up into individual markdown documents](https://github.com/Home-Is-Where-You-Hang-Your-Hack/ble_monitor/tree/jekyll/docs/_devices) with parameters in YAML headers that are used to populate templates based on criteria. Currently the only page that works like this is [supported devices by manufacture](https://home-is-where-you-hang-your-hack.github.io/ble_monitor/by_brand) where devices are grouped under manufacture.  I want to add a page where users can search by the broadcasted property such as "temperature".  The YAML headers can include any information and can query the information by whatever seems appropriate (all devices with encryption, all devices that require active scan, etc).  When a new device is added, just the device markdown file will need to be with the required header information (I'll add a template), and all of the various device queries will be populated.

**Reply by Ernst79** [reply]
Whoaw, this looks very good! I’m very impressed, looks so much better than the old (extremely long) page. Don’t forget to put yourself at the credits 😄

**Comment by Ernst79** [comment]
@Thrilleratplay I'm not sure if it is possible, but as you might have seen, I have created a pypi package (`bleparser`) that will take care of the conversion of the BLE advertisement to readable data, such that it can be used for other components as well. If possible, it would be helpful if we can create a separate page for the `bleparser`, such that we do not have to maintain two "documentations". (https://github.com/Ernst79/bleparser)

**Reply by Thrilleratplay** [reply]
@Ernst79 Of course.  I'll finish cleaning up what I currently have and update it for the newest devices.  If I do not have too many distractions, I should have that done or presentable by the end of the day.  

When reading #34, I assumed the standalone parser was going to be lower level, like aioblescan or bleson, but this makes more sense.  The way the documentation is being generated, each device can have whatever relevant data added it it but displayed in a way that is fitting (HomeAssistant user level vs dev user level).  Before creating dev level documentation, some ambiguities that exist should be clarified or at least rewritten (examples: "probably encrypted (not confirmed yet)", "we do not have accurate periodicity information yet" and many of the broadcast rates).  Again, let me finish my current restructuring of the current documentation and then it should be easier to see what is needed for developer level documentation.

**Reply by Ernst79** [reply]
For `bleparser`, I decided to only do the parsing, and I think I leave it that way. There are other tools that can do the data gathering, why doing it twice 😄 Perhaps I'll add an example later. 

About the documentation, let me know when I need to help getting this developer level info updated. There is no hurry. It's holiday season here, so I will spend a little less time after this week for a couple of weeks.

**Comment by Thrilleratplay** [comment]
@Ernst79  I enabled github pages mainly to ensure that the Jekyll config settings worked correctly for this repo.  

https://custom-components.github.io/ble_monitor/

I had difficulty getting the relative path of the image assets correct.  As I was typing this, I also remembered using `git mv` for the images, so the README.md is not display most pictures.  I'll submit a PR for the SensorPush integration in a little while.  I will also copy the images back for README.md.

I am not sure how much information to be left in README.md and info.md, I'll leave that up to you.  I will also add a page for ble_parser to begin incorporating that information.


@Magalex2x14 If you have the time, can you also review https://custom-components.github.io/ble_monitor/ to ensure there is no misinformation or missing information?

**Reply by Ernst79** [reply]
@Thrilleratplay Im working on it now, already created the SensorPush pages. But it is giving me a hard time to get it running locally in wsl2/visual studio code. It is running now, but not showing the theme correct. But I’m totally new to these Jekyll docs, and it is not really intuitive how to get it up and running. 

There is no need to put the images back, I will remove most of the content of the readme and provide links to the Jekyll docs. But first I have to get it up and running locally.

**Reply by Thrilleratplay** [reply]
If you are on a Linux machine, I been running `docker run --rm  -p 4000:4000 -v $(pwd):/site bretfisher/jekyll-serve` in the `docs` directory.  It should take care of everything.

**Reply by Ernst79** [reply]
I normally work on a Windows machine with wsl2 (Ubuntu). Not sure if this docker command works there as well, I’ll try later.

**Reply by Thrilleratplay** [reply]
Ok.  Sorry, hopefully it does as it takes care of all of dependencies.  The theme not displaying correct is likely due to it missing `jekyll-remote-theme`.  It is in the Gemfile, use `bundle install` to install all of the dependencies listed.  If `bundle` is not available, run `gem install bundle` (maybe with `sudo`, I am not sure how WSL deals with admin permissions).

**Reply by Ernst79** [reply]
Finally, it crashed on `exec jekyll serve`, but I hade to use `bundle exec jekyll serve`.

**Reply by Ernst79** [reply]
I've pushed some minor fixes to the master branch with the new SensorPush sensors. I'm getting it how it works :smile 

Do you know where I can fix the word "manufacture" (it misses an "r" at the end). Find didn't help, probably looking in the wrong place. 
![image](https://user-images.githubusercontent.com/24375847/130642532-71bcc216-15b2-4ed2-ac06-cb6fd72c4b8a.png)

**Reply by Thrilleratplay** [reply]
https://github.com/custom-components/ble_monitor/blame/master/docs/supported_devices/by_brand.md#L3

It has its quirks but hopefully will save time in the long run.

**Reply by Ernst79** [reply]
Thanks, fixed. Will look into the readme and info.md tomorrow. I think I make it very small, only a short description and a link to the docs, faq, etc.

**Reply by Ernst79** [reply]
I try to tweak the layout a bit more, as I would prefer to have a title on top of the page. I managed to get the TOC below the Title and a short text, but the original TOC is also still above the Main title. Is it possible to remove the top TOC? 

![image](https://user-images.githubusercontent.com/24375847/130675745-746e8903-d769-4497-a494-aa679d4b84c1.png)

**Reply by Ernst79** [reply]
Nevermind, figured it out, its in default.html

**Reply by Ernst79** [reply]
Made some tweaks in the layout and corrected some heading levels. Also fixed the README.md and info.md. That's it for today :smile

**Reply by Thrilleratplay** [reply]
I ran into that too, where I would want to display one header which did not need a TOC.  Use the `fake-h1` CSS class like so:

```html
<span class="fake-h1">Encrypted Advertisements</span>
```
